dataiku/dss-plugin-tableau-hyper
# Chore: Use a temporary directory rather than .cache when formatting from or exporting to the hyper file

## Overview

Card: [184434](https://app.shortcut.com/dataiku/story/175809/tableau-plugin-use-tmp-rather-than-cache-when-generating-the-hyper-file#activity-184434)

- Removed usage of `.cache` directory in favor of built-in `tempfile` methods (to create temporary directories and files).
- Used a class `TemporaryFile` to de-couple logic and make it easy to unit tests

## Run unit tests

Unit tests can be run with either (installing requirements worked on python 3.7, but NOT on python 3.9 for info):
```bash
make unit-tests

# Or specific tests from a file
pytest python-test/test_tableau_table_writer.py
```

## Manually tests the components and changes

See: [Dku doc on "Tableau Hyper Export" plugin](https://www.dataiku.com/product/plugins/tableau-hyper-export/) which explains how to use the 3 components of the plugin

#### To test the exporter from DSS dataset to Tableau Hyper file (`tableau-hyper_export`):
- Note: this exporter was not using a temporary file before, and is still not using one. We test it just to make sure it still work like before
- A first method (another method will be used for the other exporter): Select a dataset on your flow
- Go to `ACTIONS` > `Export` and stay on the tab `DOWNLOAD`
- For `Format` select `Tableau hyper extracta (.hyper)`
- Click on `DOWNLOAD`
- You can check on Tableau if the file seems correct, or use this file as the input to the formatter (see below)

#### To test the exporter upload DSS dataset as Tableau Hyper file to Tableau Server or Tableau Cloud (`tableau-hyper_upload`): 
- If you don't have an account, ask me so i can give you an API key
- Another method to test the exporter: on a flow, 
- select a dataset > `Action` > `Other recipes / Export to folder` 
- In `SET` in `Outputs` , fill in `Add folder` , using `filesystem_managed` > `Create Folder` > `CREATE RECIPE`
- On the recipe that was created: Go to `Other exports` tab
```
# Tableau Server
Authentication type: Personal Access Token preset
Tableau Server PTA Preset: <your_preset>

# Destination
Select project from list: ✅
Project: <any_project>
```
- And `SAVE` the recipe config
- Now go back on the flow, and `Run` the recipe
- Check on Tableau that the dataset has been correctly uploaded in the hyper format
- Look in the job logs for:  `Defined the temporary output file` and check that the temporary file has been deleted

	
#### To test the formatter (`hyper-format`): 
- Create a new folder and add an `.hyper` file in the folder
- In the folder, do `ACTIONS` > `Create dataset`
- Then select your hyper file (or just set `Files selection: All`) and run `TEST & GET SCHEMA`
- On the `Format / Preview` tab, select the `Type` to be an `Tableau Hyper (.hyper)`.
	- DSS will probably try to launch the formatter but there will be an error because the `Schema name`, `Table name` and `File compression` were not set
	- Set `Schema name` and `Table name` (in my example: `Extract` for both), and `File compression` (in my ex: `None`)
	- It might fail again, this was happening before this PR, and you will need to go back to the `Files` tab, then back to `Format / Preview` tab and enter again your `Schema name` and `Table name` and click on `UPDATE PREVIEW`
- Now you should be able to preview the hyper file, and you can click on `CREATE`
- A new dataset should have been created from your hyper file
- In the main logs of DSS (there is no "job" when formatting), you can look for a log: `Created temporary file to store future buffer stream from Hyper` to see the file path of the temporary file, and check that the file has been deleted
